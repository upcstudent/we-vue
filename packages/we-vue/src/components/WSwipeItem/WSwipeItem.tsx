import Vue from 'vue'

interface options extends Vue {
  $parent: Vue & {
    vertical: number
    computedWidth: number
    computedHeight: number
    swipes: Array<any>
  }
}

export default Vue.extend<options>().extend({
  name: 'w-swipe-item',

  data () {
    return {
      offset: 0,
    }
  },

  beforeCreate (): void {
    this.$parent.swipes.push(this)
  },

  destroyed (): void {
    this.$parent.swipes.splice(this.$parent.swipes.indexOf(this), 1)
  },

  render (h) {
    const { vertical, computedWidth, computedHeight } = this.$parent

    const style = {
      width: computedWidth + 'px',
      height: vertical ? computedHeight + 'px' : '100%',
      transform: vertical ? `translate3d(0, ${this.offset}px, 0)` : `translate3d(${this.offset}px, 0, 0)`,
    }

    return (
      <div class="wv-swipe-item" style={style} {...{ on: this.$listeners }}>
        {this.$slots.default}
      </div>
    )
  },
})
