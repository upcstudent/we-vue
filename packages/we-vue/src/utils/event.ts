import { isServer } from './'

type EventHandler = (event: Event) => void

export let supportsPassive = false

if (!isServer) {
  try {
    const opts = {}
    Object.defineProperty(opts, 'passive', {
      get () {
        /* istanbul ignore next */
        supportsPassive = true
      },
    })
    window.addEventListener('test-passive', null as any, opts)
  } catch (e) {}
}

export function on (target: Element | Window | Document, event: string, handler: EventHandler, passive = false) {
  !isServer &&
    target.addEventListener(
      event,
      handler,
      supportsPassive ? { capture: false, passive } : false
    )
}

export function off (target: Element | Window | Document, event: string, handler: EventHandler) {
  !isServer && target.removeEventListener(event, handler)
}
