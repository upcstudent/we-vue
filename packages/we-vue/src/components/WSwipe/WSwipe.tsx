import '../../scss/swipe.scss'

import Vue from 'vue'

// Utils
import mixins, { ExtractVue } from '../../utils/mixins'
import { off, on } from '../../utils/event'

import SwipeItem from '../WSwipeItem'

// Mixins
import Touchable from '../../mixins/touchable'

type SwipeItemInstance = InstanceType<typeof SwipeItem>

// Types
interface options extends Vue {
  timer: any
}

export default mixins<options &
  ExtractVue<[typeof Touchable]>
>(Touchable).extend({
  name: 'w-swipe',

  props: {
    width: Number,
    height: Number,
    autoplay: Number,
    vertical: Boolean,
    indicatorColor: String,
    defaultIndex: {
      type: Number,
      default: 0,
    },
    showIndicators: {
      type: Boolean,
      default: true,
    },
    duration: {
      type: Number,
      default: 500,
    },
    prevent: Boolean,
    noDragWhenSingle: {
      type: Boolean,
      default: true,
    },
    loop: {
      type: Boolean,
      default: true,
    },
    touchable: {
      type: Boolean,
      default: true,
    },
  },

  data () {
    return {
      computedWidth: 0 as number,
      computedHeight: 0 as number,
      offset: 0 as number,
      active: 0 as number,
      deltaX: 0 as number,
      deltaY: 0 as number,
      swipes: [] as Array<SwipeItemInstance>,
      swiping: false,
      rendered: false,
    }
  },

  mounted () {
    this.initialize()

    if (!this.$isServer) {
      on(window, 'resize', this.onResize, true)
    }
  },

  activated () {
    if (this.rendered) {
      this.initialize()
    }

    this.rendered = true
  },

  destroyed () {
    this.clear()

    if (!this.$isServer) {
      off(window, 'resize', this.onResize)
    }
  },

  watch: {
    swipes () {
      this.initialize()
    },

    defaultIndex () {
      this.initialize()
    },

    autoplay (autoplay) {
      if (!autoplay) {
        this.clear()
      } else {
        this.autoPlay()
      }
    },
  },

  computed: {
    count (): number {
      return this.swipes.length
    },

    delta (): number {
      return this.vertical ? this.deltaY : this.deltaX
    },

    size (): number {
      return this.vertical ? this.computedHeight : this.computedWidth
    },

    trackSize (): number {
      return this.count * this.size
    },

    isCorrectDirection (): boolean {
      const expect = this.vertical ? 'vertical' : 'horizontal'
      return this.direction === expect
    },

    trackStyle (): object {
      const mainAxis = this.vertical ? 'height' : 'width'
      const crossAxis = this.vertical ? 'width' : 'height'

      return {
        [mainAxis]: `${this.trackSize}px`,
        [crossAxis]: this[crossAxis] ? `${this[crossAxis]}px` : '',
        transitionDuration: `${this.swiping ? 0 : this.duration}ms`,
        transform: `translate${this.vertical ? 'Y' : 'X'}(${this.offset}px)`,
      }
    },

    indicatorStyle (): object {
      return {
        backgroundColor: this.indicatorColor,
      }
    },

    activeIndicator (): number {
      return (this.active + this.count) % this.count
    },
  },

  methods: {
    initialize (active?: number): void {
      active = active || this.defaultIndex

      this.clear()
      if (this.$el) {
        const rect = this.$el.getBoundingClientRect()
        this.computedWidth = this.width || rect.width
        this.computedHeight = this.height || rect.height
      }
      this.swiping = true
      this.active = active
      this.offset = this.count > 1 ? -this.size * this.active : 0
      this.swipes.forEach(swipe => {
        swipe.offset = 0
      })
      this.autoPlay()
    },

    onResize (): void {
      this.initialize(this.activeIndicator)
    },

    onTouchStart (e: TouchEvent): void {
      if (!this.touchable) return

      this.clear()
      this.swiping = true
      this.touchStart(e)
      this.correctPosition()
    },

    onTouchMove (e: TouchEvent): void {
      if (!this.touchable || !this.swiping) return

      this.touchMove(e)

      if (this.isCorrectDirection) {
        e.preventDefault()
        e.stopPropagation()
        this.move({ offset: Math.min(Math.max(this.delta, -this.size), this.size) })
      }
    },

    onTouchEnd (): void {
      if (!this.touchable || !this.swiping) return

      if (this.delta && this.isCorrectDirection) {
        const offset = this.vertical ? this.offsetY : this.offsetX

        this.move({
          pace: offset > 0 ? (this.delta > 0 ? -1 : 1) : 0,
          emitChange: true,
        })
      }

      this.swiping = false
      this.autoPlay()
    },

    move ({ pace = 0, offset = 0, emitChange }: { pace?: number, offset?: number, emitChange?: boolean}): void {
      const { delta, active, count, swipes, trackSize } = this
      const atFirst = active === 0
      const atLast = active === count - 1
      const outOfBounds =
        !this.loop &&
        ((atFirst && (offset > 0 || pace < 0)) || (atLast && (offset < 0 || pace > 0)))

      if (outOfBounds || count <= 1) {
        return
      }

      if (swipes[0]) {
        swipes[0].offset = atLast && (delta < 0 || pace > 0) ? trackSize : 0
      }

      if (swipes[count - 1]) {
        swipes[count - 1].offset = atFirst && (delta > 0 || pace < 0) ? -trackSize : 0
      }

      if (pace && active + pace >= -1 && active + pace <= count) {
        this.active += pace

        if (emitChange) {
          this.$emit('change', this.activeIndicator)
        }
      }

      this.offset = Math.round(offset - this.active * this.size)
    },

    swipeTo (index: number): void {
      this.swiping = true
      this.resetTouchStatus()
      this.correctPosition()
      setTimeout(() => {
        this.swiping = false
        this.move({
          pace: (index % this.count) - this.active,
          emitChange: true,
        })
      }, 30)
    },

    correctPosition (): void {
      if (this.active <= -1) {
        this.move({ pace: this.count })
      }
      if (this.active >= this.count) {
        this.move({ pace: -this.count })
      }
    },

    range (num: number, arr: number[]): number {
      return Math.min(Math.max(num, arr[0]), arr[1])
    },

    clear (): void {
      clearTimeout(this.timer)
    },

    autoPlay (): void {
      const { autoplay } = this
      if (autoplay && this.count > 1) {
        this.clear()
        this.timer = setTimeout(() => {
          this.swiping = true
          this.resetTouchStatus()
          this.correctPosition()

          setTimeout(() => {
            this.swiping = false
            this.move({
              pace: 1,
              emitChange: true,
            })
            this.autoPlay()
          }, 30)
        }, autoplay)
      }
    },
  },

  render (h) {
    const { count, activeIndicator } = this

    const Indicator = this.$slots.indicator || (
      this.showIndicators && count > 1 &&
      <div class="wv-swipe__indicators">
        {
          Array(...Array(count)).map((empty, index) => (
            <i
              key={index}
              class={{ 'wv-swipe__indicator--active': index === activeIndicator }}
              style={index === activeIndicator ? this.indicatorStyle : null}
            />
          ))
        }
      </div>
    )

    return (
      <div class="wv-swipe">
        <div
          style={this.trackStyle}
          class="wv-swipe__wrapper"
          ref="track"
          onTouchstart={this.onTouchStart}
          onTouchmove={this.onTouchMove}
          onTouchend={this.onTouchEnd}
          onTouchcancel={this.onTouchEnd}
        >
          {this.$slots.default}
        </div>
        {Indicator}
      </div>
    )
  },
})
