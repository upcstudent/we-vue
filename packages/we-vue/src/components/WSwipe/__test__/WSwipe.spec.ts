import { mount } from '@vue/test-utils'
import WSwipe from '../WSwipe'
import WSwipeItem from '../../WSwipeItem'
import { dragAndHoldHelper, horizontalDrag, verticalDrag } from '@/test/unit/utils'
import faker from 'faker'

describe('swipe', () => {
  test('create', () => {
    const wrapper = mount(WSwipe, {
      propsData: {},
    })

    expect(wrapper.name()).toBe('w-swipe')
    expect(wrapper.classes()).toContain('wv-swipe')

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('create with single swipe-item', () => {
    const wrapper = mount(WSwipe, {
      attachToDocument: true,
      propsData: {
        height: 120,
      },
      slots: {
        default: [WSwipeItem],
      },
    })

    expect(wrapper.findAll(WSwipeItem)).toHaveLength(1)
    expect(wrapper.vm.count).toBe(1)
  })

  test('create with swipe-items', () => {
    const wrapper = mount(WSwipe, {
      attachToDocument: true,
      propsData: {
        height: 120,
        showIndicators: true,
      },
      slots: {
        default: [WSwipeItem, WSwipeItem, WSwipeItem],
      },
    })

    expect(wrapper.vm.count).toBe(3)
    expect(wrapper.vm.activeIndicator).toBe(0)
  })

  test('drag to right', () => {
    const wrapper = mount(WSwipe, {
      attachToDocument: true,
      propsData: {
        height: 120,
      },
      slots: {
        default: [WSwipeItem, WSwipeItem, WSwipeItem],
      },
    })

    const swipeWrapperItem = wrapper.find('.wv-swipe__wrapper')

    horizontalDrag(swipeWrapperItem, 0, 100)
    expect(wrapper.vm.active).toBe(-1)
    horizontalDrag(swipeWrapperItem, 0, 100)
    expect(wrapper.vm.active).toBe(1)
    horizontalDrag(swipeWrapperItem, 0, 100)
    expect(wrapper.vm.active).toBe(0)
  })

  test('drag to left', () => {
    const wrapper = mount(WSwipe, {
      attachToDocument: true,
      propsData: {
        height: 120,
      },
      slots: {
        default: [WSwipeItem, WSwipeItem, WSwipeItem],
      },
    })

    const swipeWrapperItem = wrapper.find('.wv-swipe__wrapper')

    horizontalDrag(swipeWrapperItem, 0, -100)
    expect(wrapper.vm.active).toBe(1)
    horizontalDrag(swipeWrapperItem, 0, -100)
    expect(wrapper.vm.active).toBe(2)
    horizontalDrag(swipeWrapperItem, 0, -100)
    expect(wrapper.vm.active).toBe(3)
    horizontalDrag(swipeWrapperItem, 0, -100)
    expect(wrapper.vm.active).toBe(1)
  })

  test('drag single slide while noDragWhenSingle is true', () => {
    const wrapper = mount(WSwipe, {
      attachToDocument: true,
      propsData: {
        height: 120,
      },
      slots: {
        default: [WSwipeItem],
      },
    })

    const distance = faker.random.number()

    dragAndHoldHelper(wrapper, distance, 0)
    expect(wrapper.vm.offset).toBe(0)

    horizontalDrag(wrapper, 0, distance)
    expect(wrapper.vm.offset).toBe(0)
  })

  test('change swipes', () => {
    const initializeSpy = jest.fn()
    const wrapper = mount(WSwipe, {
      attachToDocument: true,
      propsData: {
        height: 120,
      },
      methods: {
        initialize: initializeSpy,
      },
    })

    wrapper.setData({
      swipes: [],
    })

    expect(initializeSpy).toHaveBeenCalled()
  })

  test('change defaultIndex', () => {
    const wrapper = mount(WSwipe, {
      attachToDocument: true,
      propsData: {
        height: 120,
        defaultIndex: 1,
      },
      slots: {
        default: [WSwipeItem, WSwipeItem, WSwipeItem],
      },
    })

    expect(wrapper.vm.active).toBe(1)
  })

  test('autoplay', () => {
    jest.useFakeTimers()

    const wrapper = mount(WSwipe, {
      attachToDocument: true,
      propsData: {
        height: 120,
        autoplay: 3000,
      },
      slots: {
        default: [WSwipeItem, WSwipeItem, WSwipeItem],
      },
    })

    jest.advanceTimersByTime(3030)
    expect(wrapper.vm.active).toBe(1)
    jest.advanceTimersByTime(3030)
    expect(wrapper.vm.active).toBe(2)
    jest.advanceTimersByTime(3030)
    expect(wrapper.vm.active).toBe(3)
    jest.advanceTimersByTime(3030)
    expect(wrapper.vm.active).toBe(1)

    jest.clearAllTimers()
  })

  test('preventScroll', () => {
    const wrapper = mount(WSwipe, {
      attachToDocument: true,
      propsData: {
        height: 120,
        prevent: true,
      },
      slots: {
        default: [WSwipeItem, WSwipeItem, WSwipeItem],
      },
    })

    verticalDrag(wrapper, 0, 100)
  })
})
