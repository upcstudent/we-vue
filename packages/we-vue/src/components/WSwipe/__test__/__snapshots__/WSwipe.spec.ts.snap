// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`swipe create 1`] = `
<div class="wv-swipe">
  <div class="wv-swipe__wrapper" style="width: 0px; transition-duration: 500ms; transform: translateX(0px);"></div>
</div>
`;
